sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -254,6 +254,20 @@ case class StaticInvoke(
     returnNullable: Boolean = true,
     isDeterministic: Boolean = true) extends InvokeLike {
 
+  // This additional constructor is added to keep binary compatibility after the addition of the
```

Comment:
Hm, technically all expressions under catalyst are private, and we don't maintain binary compatibility here. For the same reason, we don't run MiMa too. I believe the downstream projects can work around by reflection.

Reply:
We also made this kind of argument change at 3.2.0 too (https://github.com/apache/spark/commit/7d8181b62f17a202ba584c7bba65b61ec4724db2) without keeping binary compatibility. I would go -1 for this change - it makes less sense to keep binary compatibility for this argument specifically in the private package [which we documented](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/package.scala#L21-L22) and [we intentionally skip binary compatibility check](https://github.com/apache/spark/blob/master/project/MimaExcludes.scala#L66).

Reply:
The warning says "between minor releases" ;)

Reply:
This is an internal API, and I think it makes less sense to make some changes to keep the binary compatibility here. We should probably mention maintenance release too - note that they were all explicitly `private[sql]` before (which we removed this accessor at SPARK-16813 to make the code easier to debug easier). Such compatibility has never been guaranteed in history.

One option might be to revert https://github.com/apache/spark/pull/35243 since it is trivial up to my knowledge, and it virtually doesn't affect anything by default in Spark 3.2.1.

Reply:
I'll leave this up to the maintainers to decide whether to revert, keep this change, or break binary compatibility. I'll add the library maintainer context here though (I maintain scalapb and sparksql-scalapb). We currently don't have a way to provide users the ability to use custom types with Datasets (such as sealed trait hierarchies). To remedy that, Spark provides `Encoder` and `Decoder` which I believe are public (?), however implementing them requires `ExpressionEncoder` which quickly takes you to use catalysts expressions to do anything useful (instantiating objects, querying them, etc). Spark currently doesn't provide a general solution in this space and apparently library maintainers (myself included) dipped in the internals, and end users depend on us for this.
 
 Maintaining compatibility in the Spark/Scala ecosystem is really time consuming for maintainers - see [this](https://github.com/typelevel/frameless#versions-and-dependencies) and [this](https://scalapb.github.io/docs/sparksql/#setting-up-your-project). The need for those versions came from users noticing problems, resulting in debugging by maintainers and so on. I'd like to ask to avoid/minimize binary breakages between maintenance releases. Breaking binary compatibility on feature releases makes it hard enough. Thank you!

Reply:
I do sympathize the pain. In order to address all such problems, expressions for API (V2 expressions) are under heavy development as a long run goal. I also agree that it's probably best to avoid the changes that unnecessarily break the compatibility of private/internal API, e.g., if that does not bring significant dev overhead.

For this PR, it would look awkward and confusing (see the comments in the code): if the developers should keep the binary compatibility in the expression at `StaticInvoke` and `Invoke` or all the expressions. In addition, we should keep adding overloaded constructors, which is not ideal for private/internal API.

`Encoder` and `Decoder` are indeed public but `ExpressionEncoder` is currently not (that is under internal `catalyst` package). We do and guarantee, with binary compatibility check, and maintain the binary compatibility and backward compatibility [as documented](https://spark.apache.org/versioning-policy.html) for public API but not for internal API.

Reply:
AFAIK this is always case by case. Yes, we don't expect people to rely on private classes such as `Expression`, but the fact is many Spark libraries are already using these private classes.

The ecosystem is very important to Spark and I think we should try our best to fix binary compatibility if it does break downstream libraries. I'm +1 to this PR.

Reply:
Once we keep this compatibility, we will have to make such exceptions every time when downstream projects are broken for using our internal or private codes. If this is very significant, and a large user group is affected, maybe we should think about making it as an exception but note that this is an exception from the norm.